titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/DifferenceResolverUtils.java

```diff
@@ -228,6 +233,51 @@ public static int countActiveNotStartedTasks(EntityHolder refJobHolder, EntityHo
         return pendingTaskIds.size();
     }
 
+    // Determines unassigned IP allocations based on the current reference view
+    public static Set<String> getUnassignedIpAllocations(JobView refJobView) {
+        // Get all IP allocations from the job
+        Set<String> unassignedIpAddressIds = refJobView.getJob().getJobDescriptor().getContainer().getContainerResources().getSignedIpAddressAllocations()
+                .stream()
+                .map(signedIpAddressAllocation -> signedIpAddressAllocation.getIpAddressAllocation().getAllocationId())
+                .collect(Collectors.toCollection(ConcurrentSkipListSet::new));
+
+        // Filter out those that are assigned
+        ((List<Task>)refJobView.getTasks()).stream()
+                .map(Task.class::cast)
+                .filter(task -> !TaskState.isTerminalState(task.getStatus().getState()))
+                .map(task -> task.getTaskContext().getOrDefault(TASK_ATTRIBUTES_IP_ALLOCATION_ID, ""))
+                .forEach(unassignedIpAddressIds::remove);
+
+        return unassignedIpAddressIds;
+    }
+
+    // Copies specific task context entries from a previous task to a replacement task
+    private static Map<String, String> getTaskContextFromPreviousTask(Task previousTask) {
```

Comment:
The idea is that it would prepare whatever task context fields need to be copied forward from a previous/replacement task. Currently that it is just the one IP Allocation field. Maybe `getInitialTaskContextFromPreviousTask` would better imply that?